rfd/0182-multi-port-tcp-app-access.md

```diff
@@ -0,0 +1,496 @@
+---
+authors: Rafał Cieślak (rafal.cieslak@goteleport.com)
+state: draft
+---
+
+# RFD 182 - Multi-port TCP app access
+
+## Required Approvers
+
+* Engineering: @zmb3 && @nklaassen && (@espadolini || @smallinsky)
+
+## What
+
+Multi-port TCP app access enables cluster administrators to set up TCP apps that expose more than
+just one port to the cluster.
+
+## Why
+
+In the past, many customers expressed interest in specifying a port range for an app
+([#5669](https://github.com/gravitational/teleport/issues/5669), [#10412](https://github.com/gravitational/teleport/issues/10412)).
+Two customers asked about multi-port support immediately after seeing a demo of VNet. Some customers
+have environments where a single logical app is spread across several ports.
+
+Before VNet ([RFD 163](https://github.com/gravitational/teleport/blob/master/rfd/0163-vnet.md)),
+introducing multi-port support was limited by the fact that the end user would need to manually
+start and manage a local proxy per each port. With VNet, this restriction is gone. We can extend
+VNet and TCP app access in such a way that the end user can connect to
+`tcp-app.teleport.example.com:1234` and the connection will be forwarded to the same port number on
+the hostname specified in the app spec.
+
+## Details
+
+### UX
+
+#### Setup
+
+There's an app called waldo which accepts connections on multiple ports. Alice wants to make the app
+available in the teleport.example.com cluster. She creates a new app and includes a new field called
+`ports` in the app definition. She leaves the port number out of the `uri` field.
+
+```yaml
+apps:
+- name: "waldo"
+  uri: "tcp://localhost"
+  ports: [4080-4090, 5095]
+  labels:
+    env: prod
+```
+
+#### Usage
+
+##### Happy path
+
+###### Regular usage
+
+Bob wants to connect from his device to waldo using a CLI client called waldo-client. Bob sees in
+Teleport Connect that the app waldo.teleport.example.com supports ports 4080 and 4081, which are the
+default ports over which waldo-client connects. Bob starts VNet, opens his terminal of choice and
+points waldo-client at the app in the cluster.
+
+```
+waldo-client waldo.teleport.example.com
+```
+
+###### Local proxy
+
+Charlie needs to connect to waldo's debug service. Its API is typically available over the port
+5095. Charlie uses Linux, where VNet is not yet available. Through `tsh apps ls` they see that the
+waldo app supports the port 5095, so they start a local proxy that targets that port:
+
+```
+tsh proxy app waldo --target-port 5095 --port 5095
```

Comment:
Why don't we support kubectl portforward format?

tsh proxy app waldo --port 5095:5095 --port 6060:6060

Reply:
Oh I'll make sure to update the RFD to reflect this, though for now probably without support for multiple ports at once.

Reply:
I can also think of cases where users would still prefer to use regular `tsh proxy app` rather than VNet - no control over the DNS zone, can't run a local resolver due to system policies etc.